File: highway-pursuit-server/highway-pursuit-server/Injected/EpisodeService.cs
using HighwayPursuitServer.Data;
using HighwayPursuitServer.Server;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace HighwayPursuitServer.Injected
{
    class EpisodeService
    {
        private readonly IHookManager _hookManager;
        private bool _terminated = false;

        public EpisodeService(IHookManager hookManager)
        {
            this._hookManager = hookManager;
            this.RegisterHooks();
        }

        public void NewGame()
        {
            Reset();
        }

        public void NewLife()
        {
            Respawn(0x1);
        }

        public bool PullTerminated()
        {
            var res = _terminated;
            _terminated = false;
            return res;
        }

        #region Hooking
        private void RegisterHooks()
        {
            IntPtr resetPtr = new IntPtr(_hookManager.GetModuleBase().ToInt32() + MemoryAdresses.RESET_OFFSET);
            Reset = Marshal.GetDelegateForFunctionPointer<Reset_delegate>(resetPtr);

            IntPtr respawnPtr = new IntPtr(_hookManager.GetModuleBase().ToInt32() + MemoryAdresses.RESPAWN_OFFSET);
            Respawn = Marshal.GetDelegateForFunctionPointer<Respawn_delegate>(respawnPtr);

            IntPtr setLivesPtr = new IntPtr(_hookManager.GetModuleBase().ToInt32() + MemoryAdresses.SET_LIVES_OFFSET);
            SetLives = Marshal.GetDelegateForFunctionPointer<SetLives_delegate>(setLivesPtr);
            _hookManager.RegisterHook(setLivesPtr, new SetLives_delegate(SetLives_Hook));
        }

        #region delegates
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        delegate void Reset_delegate();

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        delegate void Respawn_delegate(byte isNotInitial);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        delegate void SetLives_delegate(byte lives);
        #endregion

        #region original function pointers
        static Reset_delegate Reset;
        static Respawn_delegate Respawn;
        static SetLives_delegate SetLives;
        #endregion

        #region hooks
        void SetLives_Hook(byte value)
        {
            SetLives(value);
            // If one life was lost
            if(value == HighwayPursuitConstants.CHEATED_CONSTANT_LIVES - 1)
            {
                _terminated = true;
            }
        }
        #endregion
        #endregion
    }
}

File: highway-pursuit-server/highway-pursuit-server/Injected/EpisodeService.cpp
#include "../pch.h"
#include "EpisodeService.hpp"
#include "MemoryAddresses.hpp"

namespace Injected
{
    // Constructor
    EpisodeService::EpisodeService(std::shared_ptr<HookManager> hookManager)
        : _hookManager(hookManager),
        _terminated(false)
    {
        RegisterHooks();
    }

    // NewGame method
    void EpisodeService::NewGame()
    {
        Reset_Base();
    }

    // NewLife method
    void EpisodeService::NewLife()
    {
        Respawn_Base(0x1);
    }

    // PullTerminated method
    bool EpisodeService::PullTerminated()
    {
        bool res = _terminated;
        _terminated = false;
        return res;
    }

    // RegisterHooks method
    void EpisodeService::RegisterHooks()
    {
        EpisodeService::Instance = this;

        // Hooked functions
        uintptr_t base = _hookManager->GetModuleBase();
        LPVOID setLivesPtr = reinterpret_cast<LPVOID>(_hookManager->GetModuleBase() + MemoryAddresses::SET_LIVES_OFFSET);
        _hookManager->RegisterHook(setLivesPtr, &SetLives_StaticHook, &SetLives_Base);

        // Other functions
        Reset_Base = (Reset_t)(base + MemoryAddresses::RESET_OFFSET);
        Respawn_Base = (Respawn_t)(base + MemoryAddresses::RESPAWN_OFFSET);
    }

    void EpisodeService::SetLives_Hook(uint8_t value)
    {
        SetLives_Base(value);
        // If one life was lost
        if (value == Data::HighwayPursuitConstants::CHEATED_CONSTANT_LIVES - 1)
        {
            _terminated = true;
        }
    }

    void __cdecl EpisodeService::SetLives_StaticHook(uint8_t lives)
    {
        if (EpisodeService::Instance != nullptr)
        {
            EpisodeService::Instance->SetLives_Hook(lives);
        }
    }

    EpisodeService* EpisodeService::Instance = nullptr;
    EpisodeService::Reset_t EpisodeService::Reset_Base = nullptr;
    EpisodeService::Respawn_t EpisodeService::Respawn_Base = nullptr;
    EpisodeService::SetLives_t EpisodeService::SetLives_Base = nullptr;
}